tools/tileGet.py

```diff
@@ -82,25 +89,48 @@ def retrieveMapServerTiles(options, west, south, east, north, decals, net):
         sx, sy = fromLatLonToTile(north, west, zoom)
         ex, ey = fromLatLonToTile(south, east, zoom)
         numTiles = (ex - sx + 1) * (ey - sy + 1)
+        
+    # opener = urllib.build_opener()
+    # opener.addheaders = [('User-agent', 'Mozilla/5.0')]
+    # urllib.install_opener(opener)
+    
+    # N = (ex - sx + 1) * (ey - sy + 1)
+    
+    if options.parallel_jobs != 0:
+        original_sigint_handler = signal.signal(signal.SIGINT, signal.SIG_IGN)
+        pool = Pool(options.parallel_jobs)
+        signal.signal(signal.SIGINT, original_sigint_handler)
+    
+    futures = []
+
+    # i = 0
     for x in range(sx, ex + 1):
         for y in range(sy, ey + 1):
             request = "%s/%s/%s/%s" % (options.url, zoom, y, x)
-#            print(request)
-#            opener = urllib.build_opener()
-#            opener.addheaders = [('User-agent', 'Mozilla/5.0')]
-#            urllib.install_opener(opener)
             filename = os.path.join(options.output_dir, "%s%s_%s.jpeg" % (options.prefix, x, y))
-            urllib.urlretrieve(request, filename)
-            if os.stat(filename).st_size < options.min_file_size:
-                raise ValueError("small file")
-            if net is not None:
-                lat, lon = fromTileToLatLon(x, y, zoom)
-                upperLeft = net.convertLonLat2XY(lon, lat)
-                lat, lon = fromTileToLatLon(x + 0.5, y + 0.5, zoom)
-                center = net.convertLonLat2XY(lon, lat)
-                print('    <decal file="%s" centerX="%s" centerY="%s" width="%s" height="%s" layer="%d"/>' %
-                      (os.path.basename(filename), center[0], center[1],
-                       2 * (center[0] - upperLeft[0]), 2 * (upperLeft[1] - center[1]), options.layer), file=decals)
+            
+            if options.parallel_jobs == 0:
+                worker(x, y, zoom, options, decals, net, request, filename)
+                # i += 1
+                # print("{%f}% ({%s})" % (100.0 * (i+1)/N, request))
+            else:
+                futures.append((
+                    x, y, pool.apply_async(worker, (options, request, filename))
+                ))
+
+    for i, (x, y, future) in enumerate(futures):
+        future.get()
+
+        # print('%f%% (%s)' % (100.0 * (i+1)/N, request))
+        if net is not None:
+        
+            lat, lon = fromTileToLatLon(x, y, zoom)
+            upperLeft = net.convertLonLat2XY(lon, lat)
+            lat, lon = fromTileToLatLon(x + 0.5, y + 0.5, zoom)
+            center = net.convertLonLat2XY(lon, lat)
+            print('    <decal file="%s" centerX="%s" centerY="%s" width="%s" height="%s" layer="%d"/>' %
+                    (os.path.basename(filename), center[0], center[1],
+                    2 * (center[0] - upperLeft[0]), 2 * (upperLeft[1] - center[1]), options.layer), file=decals)
 
```

Comment:
Does this really work for the non-parallel case? It seems to me futures will be empty then so no processing will take place

Reply:
It does. Futures are only inserted if the pool has threads. If the pool is empty, the worker code is run in line 113 without creating futures.

Tho I've just noticed a small bug in line 113 related with unnecessary arguments, will fix.

Reply:
Yes, the images will be retrieved but the decal lines will not be printed, will they?